types/react/index.d.ts

```diff
@@ -560,6 +560,164 @@ declare namespace React {
+     * React performs its DOM mutations, before sibling components have been updated. Use this to perform
+     * custom DOM mutations.
+     *
+     * Prefer the standard `useEffect` when possible to avoid blocking visual updates.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#usemutationeffect
+     */
+    function useMutationEffect(effect: EffectCallback, inputs?: InputIdentityList): void;
+    /**
+     * The signature is identical to `useEffect`, but it fires synchronously after all DOM mutations.
+     * Use this to read layout from the DOM and synchronously re-render. Updates scheduled inside
+     * `useLayoutEffect` will be flushed synchronously, before the browser has a chance to paint.
+     *
+     * Prefer the standard `useEffect` when possible to avoid blocking visual updates.
+     *
+     * If you’re migrating code from a class component, `useLayoutEffect` fires in the same phase as
+     * `componentDidMount` and `componentDidUpdate`.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#uselayouteffect
+     */
+    function useLayoutEffect(effect: EffectCallback, inputs?: InputIdentityList): void;
+    /**
+     * Accepts a function that contains imperative, possibly effectful code.
+     *
+     * @param effect Imperative function that can return a cleanup function
+     * @param inputs If present, effect will only activate if the values in the list change.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#useeffect
+     */
+    function useEffect(effect: EffectCallback, inputs?: InputIdentityList): void;
+    // NOTE: this does not accept strings, but this will have to be fixed by removing strings from type Ref<T>
+    /**
+     * `useImperativeMethods` customizes the instance value that is exposed to parent components when using
+     * `ref`. As always, imperative code using refs should be avoided in most cases.
+     *
+     * `useImperativeMethods` should be used with `React.forwardRef`.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#useimperativemethods
+     */
+    function useImperativeMethods<T, R extends T>(ref: Ref<T>|undefined, init: () => R, inputs?: InputIdentityList): void;
+    // I made 'inputs' required here and in useMemo as there's no point to memoizing without the memoization key
+    // useCallback(X) is identical to just using X, useMemo(() => Y) is identical to just using Y.
+    /**
+     * `useCallback` will return a memoized version of the callback that only changes if one of the `inputs`
+     * has changed.
+     *
+     * @version experimental
+     * @see https://reactjs.org/docs/hooks-reference.html#usecallback
+     */
+    // TODO (TypeScript 3.0): <T extends (...args: never[]) => unknown>
+    function useCallback<T extends (...args: any[]) => any>(callback: T, inputs: InputIdentityList): T;
```

Comment:
There is an inference issue with `useCallback` that creates silent implicit `any`s, but I could not find any way to work around this. Even completely removing the `extends` clause still creates silent `any`s.

```ts
<button
  onClick={React.useCallback(e => {
    // this works...
    e.preventDefault()
    // but e is actually any!
    e.nonsense
  })}
/>
```